crates/wasi-http/src/handler.rs

```diff
@@ -0,0 +1,563 @@
+                    }
+                }
+
+                // Otherwise, if no timeouts have elapsed, we set
+                // `accept_concurrent` to true and, if it wasn't already true
+                // before, poll the future one more time so it can ask for
+                // another task if appropriate.
+                if !accept_concurrent.swap(true, Relaxed) {
+                    return future.as_mut().poll(cx);
+                }
+            }
+
+            poll
+        })
+        .await?
+    }
+}
+
+impl<S, T> Drop for Worker<S, T>
+where
+    T: Send,
+    S: HandlerState<StoreData = T>,
+{
+    fn drop(&mut self) {
+        self.set_available(false);
+    }
+}
+
+/// Represents the state of a web server.
+///
+/// Note that this supports optional instance reuse, enabled when
+/// `S::max_instance_reuse_count()` returns a number greater than one.  See
+/// [`Self::push`] for details.
+pub struct ProxyHandler<S, T: 'static>(Arc<ProxyHandlerInner<S, T>>);
+
+impl<S, T> Clone for ProxyHandler<S, T> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+impl<S, T> ProxyHandler<S, T> {
+    /// Create a new `ProxyHandler` with the specified application state and
+    /// pre-instance.
+    pub fn new(state: S, instance_pre: ProxyPre<T>) -> Self {
+        Self(Arc::new(ProxyHandlerInner {
+            state,
+            instance_pre,
+            next_id: AtomicU64::from(0),
+            task_queue: Default::default(),
+            worker_count: AtomicUsize::from(0),
+        }))
+    }
+
+    /// Push a task to the task queue for this handler.
+    ///
+    /// This will either spawn a new background worker to run the task or
+    /// deliver it to an already-running worker.
+    pub fn spawn(&self, req_id: Option<u64>, task: TaskFn<T>)
+    where
+        T: Send,
+        S: HandlerState<StoreData = T>,
+    {
+        match self.0.state.max_instance_reuse_count() {
+            0 => panic!("`max_instance_reuse_count` must be at least 1"),
+            _ => {
+                if self.0.worker_count.load(Relaxed) == 0 {
+                    // There are no available workers; skip the queue and pass
+                    // the task directly to the worker, which improves
+                    // performance as measured by `wasmtime-server-rps.sh` by
+                    // about 15%.
+                    self.start_worker(Some(task), req_id);
+                } else {
+                    self.0.task_queue.push(task);
+                    // Start a new worker to handle the task if the last worker
+                    // just went unavailable.  See also `Worker::set_available`
+                    // for what happens if the available worker count goes to
+                    // zero right after we check it here, and note that we only
+                    // check the count _after_ we've pushed the task to the
+                    // queue.
+                    if self.0.worker_count.load(SeqCst) == 0 {
```

Comment:
s/SeqCst/Relaxed/

Reply:
Are you sure about that?  The ordering of this operation relative to the `Queue::push` that happens before this must be preserved, whereas `Relaxed` is documented to have "No ordering constraints, only atomic operations".  I imagine we could get away with something less strict than `SeqCst` here (maybe `Release`?), but `Relaxed` seems too weak in this case.

Reply:
In this case `worker_count` isn't protecting anything `unsafe` or anything like that so AFAIK `Relaxed` is all that's needed. Anything stronger is only required typically when `unsafe` is involved or it's specifically trying to synchronize with some other modification of `worker_count` to see updates made somewhere else, but in this case the state is in a mutex anyway so I don't think that's required

Reply:
Okay, I just pushed an update with more comments and also changed a _different_ `SeqCst` to `Relaxed`.  I'm guessing the remaining `SeqCst`s are stronger than we actually need, but the ordering requirements between the `Queue` operations involving a `Mutex` and the `worker_count` `AtomicUsize` operations are a bit subtle, and I'm erring on the side of paranoia here.